src/app-logic/web-channel.js

```diff
@@ -3,54 +3,96 @@
  */
 
-/**
- * The messages are typed as an object so that the "type" field can be extracted
- * using the $Keys utility type.
- */
-type MessageToBrowserObject = {|
-  STATUS_QUERY: {| type: 'STATUS_QUERY', requestId: number |},
-  ENABLE_MENU_BUTTON: {| type: 'ENABLE_MENU_BUTTON', requestId: number |},
+export type MessageToBrowser<Body: MessageToBrowserBody> = {|
+  requestId: number,
+|} & Body;
+
+export type MessageToBrowserBody =
+  | StatusQueryMessageToBrowserBody
+  | EnableMenuButtonMessageToBrowserBody
+  | GetProfileMessageToBrowserBody
+  | GetSymbolTableMessageToBrowserBody
+  | QuerySymbolicationApiMessageToBrowserBody;
+
+type StatusQueryMessageToBrowserBody = {| type: 'STATUS_QUERY' |};
+type EnableMenuButtonMessageToBrowserBody = {| type: 'ENABLE_MENU_BUTTON' |};
+type GetProfileMessageToBrowserBody = {| type: 'GET_PROFILE' |};
+type GetSymbolTableMessageToBrowserBody = {|
+  type: 'GET_SYMBOL_TABLE',
+  debugName: string,
+  breakpadId: string,
+|};
+type QuerySymbolicationApiMessageToBrowserBody = {|
+  type: 'QUERY_SYMBOLICATION_API',
+  path: string,
+  requestJson: string,
 |};
 
-/**
- * The messages are typed as an object so that the "type" field can be extracted
- * using the $Keys utility type.
- */
-type MessageFromBrowserObject = {|
-  STATUS_RESPONSE: {|
-    type: 'STATUS_RESPONSE',
-    menuButtonIsEnabled: boolean,
-    requestId: number,
-  |},
-  ENABLE_MENU_BUTTON_DONE: {|
-    type: 'ENABLE_MENU_BUTTON_DONE',
-    requestId: number,
-  |},
+export type MessageFromBrowser<R: ResponseFromBrowser> =
+  | OutOfBandErrorMessageFromBrowser
+  | ErrorResponseMessageFromBrowser
+  | SuccessResponseMessageFromBrowser<R>;
+
+type OutOfBandErrorMessageFromBrowser = {|
+  errno: number,
+  error: string,
+|};
+
+type ErrorResponseMessageFromBrowser = {|
+  type: 'ERROR_RESPONSE',
+  requestId: number,
+  error: string,
 |};
 
-// Extract out the different values. Exported for tests.
-export type MessageToBrowser = $Values<MessageToBrowserObject>;
-export type MessageFromBrowser = $Values<MessageFromBrowserObject>;
-export type MessageFromBrowserTypes = $Keys<MessageFromBrowserObject>;
+type SuccessResponseMessageFromBrowser<R: ResponseFromBrowser> = {
+  type: 'SUCCESS_RESPONSE',
+  requestId: number,
+  response: R,
+};
+
+export type ResponseFromBrowser =
+  | ResponseToStatusQuery
+  | ResponseToEnableMenuButton
+  | ResponseToGetProfile
+  | ResponseToGetSymbolTable
+  | ResponseToQuerySymbolicationApi;
+
+type ResponseToStatusQuery = {|
+  menuButtonIsEnabled: boolean,
+  // The version indicates which message types are supported by the browser.
+  // No version:
+  //   Shipped in Firefox 76.
+  //   Supports the following message types:
+  //    - STATUS_QUERY
+  //    - ENABLE_MENU_BUTTON
+  // Version 1:
+  //   Shipped in Firefox 93.
+  //   Adds support for the following message types:
+  //    - GET_PROFILE
+  //    - GET_SYMBOL_TABLE
+  //    - QUERY_SYMBOLICATION_API
+  version?: number,
```

Comment:
I kind of wonder if we should send `version` with every answer, that could make `_fixupOldResponseMessageIfNeeded` more explicit...

Reply:
Interesting idea!

Another thing we could do would be to have a ProfilerWebChannel class, and call `STATUS_QUERY` during the constructor. Then the message handling would always know the version.

Reply:
Yeah, this could work too. We'd need to store a promise for this initial response and `await` it in every calls, but that sounds doable.

Reply:
or maybe require the version as a constructor argument, but also have an async function that does creates the instance

Reply:
Actually, this approach would have a chicken-and-egg problem. To know the version, we need to process a `STATUS_RESPONSE` message. And in order to process that message properly, we need to know the version, to know if it needs upgrading.

I'll investigate the suggestion of putting a version everywhere.

Reply:
I gave that a try but it only made things unnecessarily complicated. I think the current approach is fine.